src/documents/models.py

```diff
@@ -314,6 +391,93 @@ def thumbnail_file(self):
     def thumbnail_url(self):
         return reverse("fetch", kwargs={"kind": "thumb", "pk": self.pk})
 
+    def set_filename(self, filename):
+        if os.path.isfile(Document.filename_to_path(filename)):
+            self.filename = filename
+
+
+def try_delete_empty_directories(directory):
+    # Go up in the directory hierarchy and try to delete all directories
+    directory = os.path.normpath(directory)
+    root = os.path.normpath(Document.filename_to_path(""))
+
+    while directory != root:
+        # Try to delete the current directory
+        try:
+            os.rmdir(directory)
+        except os.error:
+            # Directory not empty, no need to go further up
+            return
+
+        # Cut off actual directory and go one level up
+        directory, _ = os.path.split(directory)
+        directory = os.path.normpath(directory)
+
+
+@receiver(models.signals.m2m_changed, sender=Document.tags.through)
+@receiver(models.signals.post_save, sender=Document)
+def update_filename(sender, instance, **kwargs):
+    # Skip if document has not been saved yet
+    if instance.filename is None:
+        return
+
+    # Build the new filename
+    new_filename = instance.generate_source_filename()
+
+    # If the filename is the same, then nothing needs to be done
+    if instance.filename == new_filename:
+        return
+
+    # Determine the full "target" path
+    path_new = instance.filename_to_path(new_filename)
+    dir_new = instance.filename_to_path(os.path.dirname(new_filename))
+
+    # Create new path
+    instance.create_source_directory()
+
+    # Determine the full "current" path
+    path_current = instance.filename_to_path(instance.filename)
+
+    # Move file
+    try:
+        os.rename(path_current, path_new)
+    except PermissionError:
+        # Do not update filename in object
+        return
+
```

Comment:
What if we get killed right here? `instance.filename` is still the old one, but the file is already at the new filename. Can we handle that somehow? Because in the next run, it will just fail in the `os.rename()` above, right?

In general, if something else is accessing the files, it might - by accident - move them around. I think we should not crash then. It raises a `FileNotFoundError`, which we need to handle some way here.

Reply:
Valid feedback! I have addressed this through change in source_filename. As we always append the document id to all filenames, we habe a fair change to find documents in case the rename failed.

What we cannot cover is someone just renaming on the filesystem level and then removing the document if suffix.